File: lib/email-compliance.ts
/**
 * Email compliance — non-negotiable for any cold-outbound product.
 *
 * Every sent email MUST carry: a working one-click unsubscribe link, a
 * physical sender address (CAN-SPAM), and an honest from/subject. This
 * module owns the unsubscribe token scheme + footer injection +
 * suppression checks.
 *
 * Token = base64url(recipientId.userId).hmac — verifiable without a DB
 * lookup, and the handler suppresses by user+email so an unsub from any
 * campaign suppresses globally for that sender.
 */

import crypto from "node:crypto"

function secret(): string {
  return process.env.UNSUB_SECRET ?? "dev-unsub-secret"
}

export function sha256Email(email: string): string {
  return crypto.createHash("sha256").update(email.trim().toLowerCase()).digest("hex")
}

export function makeUnsubToken(recipientId: string, userId: string): string {
  const payload = Buffer.from(`${recipientId}.${userId}`).toString("base64url")
  const sig = crypto.createHmac("sha256", secret()).update(payload).digest("base64url")
  return `${payload}.${sig}`
}

export function verifyUnsubToken(
  token: string,
): { recipientId: string; userId: string } | null {
  const [payload, sig] = token.split(".")
  if (!payload || !sig) return null
  const expected = crypto
    .createHmac("sha256", secret())
    .update(payload)
    .digest("base64url")
  // Constant-time comparison.
  if (
    sig.length !== expected.length ||
    !crypto.timingSafeEqual(Buffer.from(sig), Buffer.from(expected))
  ) {
    return null
  }
  const decoded = Buffer.from(payload, "base64url").toString("utf-8")
  const [recipientId, userId] = decoded.split(".")
  if (!recipientId || !userId) return null
  return { recipientId, userId }
}

/**
 * Append the legally-required footer to an email body.
 * physicalAddress is captured at mailbox-connection time.
 */
export function appendComplianceFooter(opts: {
  body: string
  unsubToken: string
  physicalAddress?: string | null
  appUrl: string
}): string {
  const unsubUrl = `${opts.appUrl}/u/${opts.unsubToken}`
  const addr = opts.physicalAddress?.trim()
  const footerLines = [
    "",
    "—",
    `Don't want these emails? Unsubscribe: ${unsubUrl}`,
  ]
  if (addr) footerLines.push(addr)
  return `${opts.body}\n${footerLines.join("\n")}`
}

File: backend-python/leadgen_backend/email_compliance.py
import hmac
import hashlib
import base64
import os

def get_secret() -> str:
    return os.getenv("UNSUB_SECRET", "dev-unsub-secret")

def sha256_email(email: str) -> str:
    return hashlib.sha256(email.strip().lower().encode("utf-8")).hexdigest()

def make_unsub_token(recipient_id: str, user_id: str) -> str:
    payload = base64.urlsafe_b64encode(f"{recipient_id}.{user_id}".encode("utf-8")).decode("ascii").rstrip("=")
    sig = base64.urlsafe_b64encode(
        hmac.new(get_secret().encode("utf-8"), payload.encode("utf-8"), hashlib.sha256).digest()
    ).decode("ascii").rstrip("=")
    return f"{payload}.{sig}"

def verify_unsub_token(token: str) -> dict[str, str] | None:
    parts = token.split(".")
    if len(parts) != 2:
        return None
    payload, sig = parts
    expected = base64.urlsafe_b64encode(
        hmac.new(get_secret().encode("utf-8"), payload.encode("utf-8"), hashlib.sha256).digest()
    ).decode("ascii").rstrip("=")
    
    if not hmac.compare_digest(sig, expected):
        return None
        
    try:
        # Add padding back if necessary
        padded_payload = payload + "=" * (-len(payload) % 4)
        decoded = base64.urlsafe_b64decode(padded_payload).decode("utf-8")
        recipient_id, user_id = decoded.split(".")
        if not recipient_id or not user_id:
            return None
        return {"recipientId": recipient_id, "userId": user_id}
    except Exception:
        return None

def append_compliance_footer(body: str, unsub_token: str, physical_address: str | None, app_url: str) -> str:
    unsub_url = f"{app_url}/u/{unsub_token}"
    addr = physical_address.strip() if physical_address else None
    
    footer_lines = [
        "",
        "—",
        f"Don't want these emails? Unsubscribe: {unsub_url}"
    ]
    if addr:
        footer_lines.append(addr)
        
    return body + "\n" + "\n".join(footer_lines)
